### PR TITLE
Fix items from losing functionality when manually inserted and/or providing player light when used remotely

### LINKED ISSUE #68075: inserting timekeeping devices into a container doesn't enable the time display
### Describe the bug

when inserting into a container from the inventory menu, or moving between containers from the inventory organization menu, a timekeeping device won't enable its timekeeping function. if it was already in your inventory, and you insert it into another container, the time display will disappear. if you drop the item or the container and pick it back up, timekeeping will be reenabled.

the calorie counter function of a smartphone for example also gets disabled from this.

### Attach save file

[Plattsmouth-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/12567835/Plattsmouth-trimmed.tar.gz)

### Steps to reproduce

spawn a smartphone and any container
open the container's inventory menu and use the insert option OR use the inventory organization screen to move the smartphone into the container
see that precise timekeeping is disabled (if no other timekeeping devices in inventory) and precise calorie counts are disabled in the eat menu

### Expected behavior

bug reverted to normal (i don't think i've ever had this problem in older versions)

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.22621.2134 (22H2)
- Game Version: 73fd431 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Kamayana**
#### Summary
Bugfixes "Fix items from losing functionality when manually inserted and/or providing player light when used remotely"

#### Purpose of change
Fixes #68075, #68435, and the first part of #69140.

Also adds the function `carrier()` to item_location, which can be used to easily get the character that's holding an item_location.

#### Describe the solution
Make `insert_item` functions return `ret_val<item *>` so they can return a pointer to the item that was added. Use this to make that newly inserted item call `on_pickup`, which adds it to the now-carrying-character's caches. This is done in `insert_item_activity_actor::finish`, `Character::store`, `item::fill_with`, and `item_pocket::fill_with`. Items with charges should now work as well.

To fix the remote lantern bug, `item::convert` will now check if the provided character is actually holding the item before adding it to the cache.

#### Describe alternatives you've considered
I originally simplified `item_location::held_by` to use the new function like this:
```
bool item_location::held_by( Character const &who ) const
{
    return carrier() == &who;
}
```
However, this made the game start failing tests in the `npc_shopkeeper_item_groups` test case. I spent a while trying to determine why, but I couldn't find anything concrete. I think it's possible those tests are relying on some sort of buggy interaction with the creature_tracker. I decided I wasn't making progress and reverted the change.

#### Testing

1. Create a smartphone and a smartphone case, with both in the inventory. Time should be visible.
2. Do the following actions individually and take note that the time will now be correctly shown/hidden as appropriate.
3. Use the insertion menu to put the phone in the case.
4. Use the AIM to insert the phone.
5. Do the previous two again, but with the phone starting outside your inventory on the ground. The clock will properly activate.
6. Use the inventory Organize Contents menu to insert the phone.
7. Click and drag the phone between containers in the inventory menu.
8. Add the "WATCH" flag to an item with charges (I used marbles). Repeat the previous steps but with inserting varying numbers of that item into a held container. The clock should activate and stay active, showing that stack merging doesn't break it.
9. Follow the steps written out in `#68075, #69140, and #68435` and see that they are now fixed.

I also ran the test suite locally and got all green results.

#### Additional context
I actually fixed the remote light not illuminating bug too, but I'll make a separate PR for that since it turns out it's not actually related.

### CODE REVIEW

**Comment by andrei8l** [comment]
> I think it's possible those tests are relying on some sort of buggy interaction with the creature_tracker. I decided I wasn't making progress and reverted the change.

Yes, the test uses a fake `item_location` for which `held_by()` always fails.
<details>
<summary>This should fix the test and let you simplify held_by</summary>

```diff
diff --git a/tests/npc_shopkeeper_item_groups_test.cpp b/tests/npc_shopkeeper_item_groups_test.cpp
index da94d66592..78f1c34cf2 100644
--- a/tests/npc_shopkeeper_item_groups_test.cpp
+++ b/tests/npc_shopkeeper_item_groups_test.cpp
@@ -39,7 +39,7 @@ TEST_CASE( "npc_shopkeeper_item_groups", "[npc][trade]" )
             std::pair<bool, bool> har_pants = has_and_can_restock( guy, pants );
             REQUIRE( har_pants.first == true );
             REQUIRE( har_pants.second == true );
-            REQUIRE( guy.wants_to_sell( { guy, &pants } ) );
+            REQUIRE( guy.wants_to_sell( { map_cursor{ tripoint_zero }, &pants } ) );
         }
     }
 
@@ -79,7 +79,7 @@ TEST_CASE( "npc_shopkeeper_item_groups", "[npc][trade]" )
             THEN( "item is available for restocking but not selling" ) {
                 REQUIRE( har_hammer.first == true );
                 REQUIRE( har_hammer.second == true );
-                REQUIRE( !guy.wants_to_sell( {guy, &hammer } ) );
+                REQUIRE( !guy.wants_to_sell( { map_cursor{ tripoint_zero }, &hammer } ) );
             }
         }
         WHEN( "condition met" ) {
@@ -88,7 +88,7 @@ TEST_CASE( "npc_shopkeeper_item_groups", "[npc][trade]" )
             THEN( "item is available for selling and restocking" ) {
                 REQUIRE( har_hammer.first == true );
                 REQUIRE( har_hammer.second == true );
-                REQUIRE( guy.wants_to_sell( { guy, &hammer } ) );
+                REQUIRE( guy.wants_to_sell( { map_cursor{ tripoint_zero }, &hammer } ) );
             }
         }
     }
@@ -101,7 +101,7 @@ TEST_CASE( "npc_shopkeeper_item_groups", "[npc][trade]" )
             THEN( "item is not available for selling or restocking" ) {
                 REQUIRE( har_multitool.first == true );
                 REQUIRE( har_multitool.second == false );
-                REQUIRE( !guy.wants_to_sell( { guy, &multitool } ) );
+                REQUIRE( !guy.wants_to_sell( { map_cursor{ tripoint_zero }, &multitool } ) );
             }
         }
         WHEN( "condition met" ) {
@@ -110,7 +110,7 @@ TEST_CASE( "npc_shopkeeper_item_groups", "[npc][trade]" )
             THEN( "item is available for selling and restocking" ) {
                 REQUIRE( har_multitool.first == true );
                 REQUIRE( har_multitool.second == true );
-                REQUIRE( guy.wants_to_sell( {guy, &multitool } ) );
+                REQUIRE( guy.wants_to_sell( { map_cursor{ tripoint_zero }, &multitool } ) );
             }
         }
     }
```

we're trading one fake `item_location` for another but this one should be benign.

</details>

**Comment by Kamayana** [comment]
> > I think it's possible those tests are relying on some sort of buggy interaction with the creature_tracker. I decided I wasn't making progress and reverted the change.
> 
> Yes, the test uses a fake `item_location` for which `held_by()` always fails.
> This should fix the test and let you simplify held_by

That worked, thank you!

File: src/item_pocket.cpp [hunk 1]
```diff
@@ -2154,23 +2179,28 @@ std::list<item> &item_pocket::edit_contents()
...
+    item *inserted = nullptr;
+    if( !into_bottom ) {
+        contents.push_front( it );
+        inserted = &contents.front();
+    } else {
+        contents.push_back( it );
+        inserted = &contents.back();
+    }
+    if( restack_charges ) {
+        restack( inserted );
```
Comment (andrei8l) [inline src/item_pocket.cpp:2201]: `inserted` can get stacked into another item if it has charges (ex: milk). This should probably be `inserted = restack( inserted );`

<details>
<summary>ASan</summary>

```gdb
==1041922==ERROR: AddressSanitizer: heap-use-after-free on address 0x616002515798 at pc 0x563e6dc2f283 bp 0x7ffc99938d10 sp 0x7ffc99938d08
READ of size 4 at 0x616002515798 thread T0
    #0 0x563e6dc2f282 in string_id<faction>::operator==(string_id<faction> const&) const /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/string_id.h:239:20
    #1 0x563e6dc2f282 in string_id<faction>::is_null() const /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/string_id.h:323:16
    #2 0x563e6dc2f282 in item::validate_ownership() const /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item.cpp:2035:20
    #3 0x563e6dc2c44b in item::get_owner() const /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item.cpp:2023:5
    #4 0x563e6dc2bcf1 in item::is_owned_by(Character const&, bool) const /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item.cpp:1947:9
    #5 0x563e6dcf6321 in item::handle_pickup_ownership(Character&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item.cpp:6371:9
    #6 0x563e6dc24a94 in item::on_pickup(Character&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item.cpp:6433:9
    #7 0x563e6dc10516 in item::fill_with(item const&, int, bool, bool, bool, bool, Character*) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item.cpp:12168:29
    #8 0x563e6b1178ec in try_insert(item_location&, std::__1::pair<item_location, int>&, int*, Character*) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/activity_actor.cpp:4321:31
    #9 0x563e6b113ebb in insert_item_activity_actor::finish(player_activity&, Character&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/activity_actor.cpp:4359:19
    #10 0x563e70b78e61 in player_activity::do_turn(Character&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_activity.cpp:376:20
    #11 0x563e6cd16d74 in do_turn() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/do_turn.cpp:484:20
    #12 0x563e6aed1824 in main /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/main.cpp:797:17
    #13 0x7fee57a4fccf  (/usr/lib/libc.so.6+0x27ccf) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
    #14 0x7fee57a4fd89 in __libc_start_main (/usr/lib/libc.so.6+0x27d89) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)
    #15 0x563e6ad93d44 in _start (/usr/src/cataclysm-dda-git/src/build_asan/cataclysm-tiles+0x91d5d44) (BuildId: b17fb25669305e6a7fd8f3d7c8d40c2490f752c2)

0x616002515798 is located 536 bytes inside of 584-byte region [0x616002515580,0x6160025157c8)
freed by thread T0 here:
    #0 0x563e6aecac4a in operator delete(void*) (/usr/src/cataclysm-dda-git/src/build_asan/cataclysm-tiles+0x930cc4a) (BuildId: b17fb25669305e6a7fd8f3d7c8d40c2490f752c2)
    #1 0x563e6bf81a59 in void std::__1::__libcpp_operator_delete[abi:v160006]<void*>(void*) /usr/bin/../include/c++/v1/new:276:3
    #2 0x563e6bf81a59 in void std::__1::__do_deallocate_handle_size[abi:v160006]<>(void*, unsigned long) /usr/bin/../include/c++/v1/new:300:10
    #3 0x563e6bf81a59 in std::__1::__libcpp_deallocate[abi:v160006](void*, unsigned long, unsigned long) /usr/bin/../include/c++/v1/new:316:14
    #4 0x563e6bf81a59 in std::__1::allocator<std::__1::__list_node<item, void*>>::deallocate[abi:v160006](std::__1::__list_node<item, void*>*, unsigned long) /usr/bin/../include/c++/v1/__memory/allocator.h:131:13
    #5 0x563e6bf81a59 in std::__1::allocator_traits<std::__1::allocator<std::__1::__list_node<item, void*>>>::deallocate[abi:v160006](std::__1::allocator<std::__1::__list_node<item, void*>>&, std::__1::__list_node<item, void*>*, unsigned long) /usr/bin/../include/c++/v1/__memory/allocator_traits.h:288:13
    #6 0x563e6bf81a59 in std::__1::list<item, std::__1::allocator<item>>::erase(std::__1::__list_const_iterator<item, void*>) /usr/bin/../include/c++/v1/list:1730:5
    #7 0x563e6e22a5a0 in item_pocket::restack(item*) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item_pocket.cpp:332:39
    #8 0x563e6e25c9f9 in item_pocket::insert_item(item const&, bool, bool, bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item_pocket.cpp:2201:9
    #9 0x563e6dc10188 in item::fill_with(item const&, int, bool, bool, bool, bool, Character*) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item.cpp:12149:42
    #10 0x563e6b1178ec in try_insert(item_location&, std::__1::pair<item_location, int>&, int*, Character*) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/activity_actor.cpp:4321:31
    #11 0x563e6b113ebb in insert_item_activity_actor::finish(player_activity&, Character&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/activity_actor.cpp:4359:19
    #12 0x563e70b78e61 in player_activity::do_turn(Character&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_activity.cpp:376:20
    #13 0x563e6cd16d74 in do_turn() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/do_turn.cpp:484:20
    #14 0x563e6aed1824 in main /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/main.cpp:797:17
    #15 0x7fee57a4fccf  (/usr/lib/libc.so.6+0x27ccf) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)

previously allocated by thread T0 here:
    #0 0x563e6aeca1b2 in operator new(unsigned long) (/usr/src/cataclysm-dda-git/src/build_asan/cataclysm-tiles+0x930c1b2) (BuildId: b17fb25669305e6a7fd8f3d7c8d40c2490f752c2)
    #1 0x563e6b1d3f33 in void* std::__1::__libcpp_operator_new[abi:v160006]<unsigned long>(unsigned long) /usr/bin/../include/c++/v1/new:266:10
    #2 0x563e6b1d3f33 in std::__1::__libcpp_allocate[abi:v160006](unsigned long, unsigned long) /usr/bin/../include/c++/v1/new:292:10
    #3 0x563e6b1d3f33 in std::__1::allocator<std::__1::__list_node<item, void*>>::allocate[abi:v160006](unsigned long) /usr/bin/../include/c++/v1/__memory/allocator.h:115:38
    #4 0x563e6b1d3f33 in std::__1::allocator_traits<std::__1::allocator<std::__1::__list_node<item, void*>>>::allocate[abi:v160006](std::__1::allocator<std::__1::__list_node<item, void*>>&, unsigned long) /usr/bin/../include/c++/v1/__memory/allocator_traits.h:268:20
    #5 0x563e6b1d3f33 in std::__1::list<item, std::__1::allocator<item>>::__allocate_node[abi:v160006](std::__1::allocator<std::__1::__list_node<item, void*>>&) /usr/bin/../include/c++/v1/list:1097:28
    #6 0x563e6b1d3f33 in std::__1::list<item, std::__1::allocator<item>>::push_back(item const&) /usr/bin/../include/c++/v1/list:1531:29
    #7 0x563e6e25c92b in item_pocket::insert_item(item const&, bool, bool, bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item_pocket.cpp:2197:18
    #8 0x563e6dc10188 in item::fill_with(item const&, int, bool, bool, bool, bool, Character*) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/item.cpp:12149:42
    #9 0x563e6b1178ec in try_insert(item_location&, std::__1::pair<item_location, int>&, int*, Character*) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/activity_actor.cpp:4321:31
    #10 0x563e6b113ebb in insert_item_activity_actor::finish(player_activity&, Character&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/activity_actor.cpp:4359:19
    #11 0x563e70b78e61 in player_activity::do_turn(Character&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_activity.cpp:376:20
    #12 0x563e6cd16d74 in do_turn() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/do_turn.cpp:484:20
    #13 0x563e6aed1824 in main /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/main.cpp:797:17
    #14 0x7fee57a4fccf  (/usr/lib/libc.so.6+0x27ccf) (BuildId: 8bfe03f6bf9b6a6e2591babd0bbc266837d8f658)

SUMMARY: AddressSanitizer: heap-use-after-free /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/string_id.h:239:20 in string_id<faction>::operator==(string_id<faction> const&) const
Shadow bytes around the buggy address:
  0x616002515500: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x616002515580: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x616002515600: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x616002515680: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x616002515700: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x616002515780: fd fd fd[fd]fd fd fd fd fd fa fa fa fa fa fa fa
  0x616002515800: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x616002515880: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x616002515900: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x616002515980: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x616002515a00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1041922==ABORTING

```

</details>

Comment (Kamayana) [reply src/item_pocket.cpp:2201]: Good call, do you have a save I could load to make sure that fixes it?

Comment (andrei8l) [reply src/item_pocket.cpp:2201]: [Tunkhannock-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/13347400/Tunkhannock-trimmed.tar.gz) insert more milk in the held container

Comment (Kamayana) [reply src/item_pocket.cpp:2201]: Thanks! Your suggestion fixed it, I'll set that PR up.

**Pull request merged**